src/dispatching/dispatcher.rs

```diff
@@ -227,150 +244,335 @@ where
+                    &self.shipping_queries_queue,
+                    query,
+                    "UpdateKind::ShippingQuery",
+                ),
+                UpdateKind::PreCheckoutQuery(query) => send(
+                    &self.requester,
+                    &self.pre_checkout_queries_queue,
+                    query,
+                    "UpdateKind::PreCheckoutQuery",
+                ),
+                UpdateKind::Poll(poll) => {
+                    send(&self.requester, &self.polls_queue, poll, "UpdateKind::Poll")
+                }
+                UpdateKind::PollAnswer(answer) => send(
+                    &self.requester,
+                    &self.poll_answers_queue,
+                    answer,
+                    "UpdateKind::PollAnswer",
+                ),
+                UpdateKind::MyChatMember(chat_member_updated) => send(
+                    &self.requester,
+                    &self.my_chat_members_queue,
+                    chat_member_updated,
+                    "UpdateKind::MyChatMember",
+                ),
+                UpdateKind::ChatMember(chat_member_updated) => send(
+                    &self.requester,
+                    &self.chat_members_queue,
+                    chat_member_updated,
+                    "UpdateKind::MyChatMember",
+                ),
+            }
+        }
+    }
+
+    async fn wait_for_handlers(&mut self) {
+        log::debug!("Waiting for handlers to finish");
+
+        // Drop all senders, so handlers can stop
+        self.messages_queue.take();
+        self.edited_messages_queue.take();
+        self.channel_posts_queue.take();
+        self.edited_channel_posts_queue.take();
+        self.inline_queries_queue.take();
+        self.chosen_inline_results_queue.take();
+        self.callback_queries_queue.take();
+        self.shipping_queries_queue.take();
+        self.pre_checkout_queries_queue.take();
+        self.polls_queue.take();
+        self.poll_answers_queue.take();
+        self.my_chat_members_queue.take();
+        self.chat_members_queue.take();
+
+        // Wait untill all handlers finish
+        self.running_handlers.by_ref().for_each(|_| async {}).await;
+    }
+}
+
+/// A token which can be used to shutdown dispatcher.
+#[derive(Clone)]
+pub struct ShutdownToken {
+    shutdown_state: Arc<AtomicShutdownState>,
+    shutdown_notify_back: Arc<Notify>,
+}
+
+impl ShutdownToken {
+    /// Tries to shutdown dispatching.
+    ///
+    /// Returns error if this dispather isn't dispatching at the moment.
+    ///
+    /// If you don't need to wait for shutdown, returned future can be ignored.
+    pub fn shutdown(&self) -> Result<impl Future<Output = ()> + '_, ShutdownError> {
+        shutdown_inner(&self.shutdown_state)
+            .map(|()| async move { self.shutdown_notify_back.notified().await })
+    }
+}
+
+/// Error occured while trying to shutdown dispatcher.
+#[derive(Debug)]
+pub enum ShutdownError {
```

Comment:
Consider using the type-state pattern instead of returning a runtime error: `pub struct Dispatcher<R, State=DispatcherStateIdle>`, etc.